// @earendil-works/gondolin/dist/src/vfs/node/loader.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
const primordials_1 = __importDefault(require("./primordials"));
const internal_binding_1 = __importDefault(require("./internal-binding"));
const internal_modules_1 = __importDefault(require("./internal-modules"));
const cache = new Map();
// node-vfs is located at host/vendor/node-vfs
// Handle both source (src/vfs/node) and compiled (dist/src/vfs/node) paths
function findVendorPath() {
    // Try from compiled location first (dist/src/vfs/node -> host/vendor)
    const fromDist = node_path_1.default.resolve(__dirname, '..', '..', '..', '..', 'vendor', 'node-vfs', 'lib', 'internal', 'vfs');
    if (node_fs_1.default.existsSync(fromDist)) {
        return fromDist;
    }
    // Fall back to source location (src/vfs/node -> host/vendor)
    const fromSrc = node_path_1.default.resolve(__dirname, '..', '..', '..', 'vendor', 'node-vfs', 'lib', 'internal', 'vfs');
    return fromSrc;
}
const VENDOR_VFS_PATH = findVendorPath();
function createRequire(parentPath) {
    return function vfsRequire(id) {
        if (id.startsWith('internal/vfs/')) {
            const modulePath = node_path_1.default.join(VENDOR_VFS_PATH, id.slice('internal/vfs/'.length) + '.js');
            return loadModule(modulePath);
        }
        if (id.startsWith('internal/')) {
            const mod = internal_modules_1.default[id];
            if (mod)
                return mod;
            throw new Error(`Unknown internal module: ${id}`);
        }
        return require(id);
    };
}
function loadModule(modulePath) {
    if (cache.has(modulePath)) {
        return cache.get(modulePath).exports;
    }
    const code = node_fs_1.default.readFileSync(modulePath, 'utf8');
    const mod = { exports: {} };
    cache.set(modulePath, mod);
    const wrapped = `(function(exports, require, module, __filename, __dirname, primordials, internalBinding) {\n${code}\n})`;
    const fn = eval(wrapped);
    const moduleDir = node_path_1.default.dirname(modulePath);
    fn(mod.exports, createRequire(modulePath), mod, modulePath, moduleDir, primordials_1.default, internal_binding_1.default);
    return mod.exports;
}
function load(name) {
    return loadModule(node_path_1.default.join(VENDOR_VFS_PATH, name + '.js'));
}
function loadProvider(name) {
    return loadModule(node_path_1.default.join(VENDOR_VFS_PATH, 'providers', name + '.js'));
}
exports.default = { load, loadProvider };
