// @earendil-works/gondolin/dist/src/vfs/node/primordials.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
// Primordials shim - frozen built-in methods used by Node.js internals
// In userland this is just a passthrough to regular built-ins
const SymbolDispose = Symbol.dispose ?? Symbol.for('nodejs.dispose');
const primordials = {
    ArrayPrototypeIndexOf: (arr, v) => arr.indexOf(v),
    ArrayPrototypePush: (arr, ...v) => {
        for (const item of v) {
            arr.push(item);
        }
        return arr.length;
    },
    ArrayPrototypeSplice: (arr, ...v) => Array.prototype.splice.apply(arr, v),
    Boolean,
    DateNow: Date.now,
    ErrorCaptureStackTrace: Error.captureStackTrace?.bind(Error) ?? (() => { }),
    Float64Array,
    FunctionPrototypeCall: (fn, thisArg, ...args) => fn.call(thisArg, ...args),
    MathCeil: Math.ceil,
    MathFloor: Math.floor,
    MathMin: Math.min,
    ObjectDefineProperties: Object.defineProperties,
    ObjectDefineProperty: Object.defineProperty,
    ObjectFreeze: Object.freeze,
    Promise,
    PromiseResolve: Promise.resolve.bind(Promise),
    SafeMap: Map,
    SafeSet: Set,
    StringPrototypeEndsWith: (s, v) => s.endsWith(v),
    StringPrototypeLastIndexOf: (s, v) => s.lastIndexOf(v),
    StringPrototypeReplaceAll: (s, a, b) => s.split(a).join(b),
    StringPrototypeSlice: (s, a, b) => s.slice(a, b),
    StringPrototypeSplit: (s, d) => s.split(d),
    StringPrototypeStartsWith: (s, v) => s.startsWith(v),
    Symbol,
    SymbolAsyncIterator: Symbol.asyncIterator,
    SymbolDispose,
};
exports.default = primordials;
