// @earendil-works/gondolin/dist/src/index.js
"use strict";
/**
 * @earendil-works/gondolin
 *
 * Alpine Linux sandbox for running untrusted code with controlled
 * filesystem and network access.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasGuestAssets = exports.getAssetDirectory = exports.getAssetVersion = exports.ensureGuestAssets = exports.HttpRequestBlockedError = exports.createHttpHooks = exports.MAX_RPC_DATA = exports.FsRpcService = exports.RpcFileHandle = exports.RpcFsBackend = exports.FsRpcClient = exports.ReadonlyProvider = exports.SandboxVfsProvider = exports.RealFSProvider = exports.MemoryProvider = exports.VirtualProvider = exports.VirtualFileSystem = exports.createVfs = exports.resolveSandboxWsServerOptionsAsync = exports.resolveSandboxWsServerOptions = exports.SandboxWsServer = exports.VM = void 0;
// Main VM interface
var vm_1 = require("./vm");
Object.defineProperty(exports, "VM", { enumerable: true, get: function () { return vm_1.VM; } });
// Server for running the sandbox
var sandbox_ws_server_1 = require("./sandbox-ws-server");
Object.defineProperty(exports, "SandboxWsServer", { enumerable: true, get: function () { return sandbox_ws_server_1.SandboxWsServer; } });
Object.defineProperty(exports, "resolveSandboxWsServerOptions", { enumerable: true, get: function () { return sandbox_ws_server_1.resolveSandboxWsServerOptions; } });
Object.defineProperty(exports, "resolveSandboxWsServerOptionsAsync", { enumerable: true, get: function () { return sandbox_ws_server_1.resolveSandboxWsServerOptionsAsync; } });
// VFS (Virtual File System) providers
var vfs_1 = require("./vfs");
Object.defineProperty(exports, "createVfs", { enumerable: true, get: function () { return vfs_1.create; } });
Object.defineProperty(exports, "VirtualFileSystem", { enumerable: true, get: function () { return vfs_1.VirtualFileSystem; } });
Object.defineProperty(exports, "VirtualProvider", { enumerable: true, get: function () { return vfs_1.VirtualProvider; } });
Object.defineProperty(exports, "MemoryProvider", { enumerable: true, get: function () { return vfs_1.MemoryProvider; } });
Object.defineProperty(exports, "RealFSProvider", { enumerable: true, get: function () { return vfs_1.RealFSProvider; } });
Object.defineProperty(exports, "SandboxVfsProvider", { enumerable: true, get: function () { return vfs_1.SandboxVfsProvider; } });
Object.defineProperty(exports, "ReadonlyProvider", { enumerable: true, get: function () { return vfs_1.ReadonlyProvider; } });
Object.defineProperty(exports, "FsRpcClient", { enumerable: true, get: function () { return vfs_1.FsRpcClient; } });
Object.defineProperty(exports, "RpcFsBackend", { enumerable: true, get: function () { return vfs_1.RpcFsBackend; } });
Object.defineProperty(exports, "RpcFileHandle", { enumerable: true, get: function () { return vfs_1.RpcFileHandle; } });
Object.defineProperty(exports, "FsRpcService", { enumerable: true, get: function () { return vfs_1.FsRpcService; } });
Object.defineProperty(exports, "MAX_RPC_DATA", { enumerable: true, get: function () { return vfs_1.MAX_RPC_DATA; } });
// HTTP hooks for network policy
var http_hooks_1 = require("./http-hooks");
Object.defineProperty(exports, "createHttpHooks", { enumerable: true, get: function () { return http_hooks_1.createHttpHooks; } });
var qemu_net_1 = require("./qemu-net");
Object.defineProperty(exports, "HttpRequestBlockedError", { enumerable: true, get: function () { return qemu_net_1.HttpRequestBlockedError; } });
// Asset management
var assets_1 = require("./assets");
Object.defineProperty(exports, "ensureGuestAssets", { enumerable: true, get: function () { return assets_1.ensureGuestAssets; } });
Object.defineProperty(exports, "getAssetVersion", { enumerable: true, get: function () { return assets_1.getAssetVersion; } });
Object.defineProperty(exports, "getAssetDirectory", { enumerable: true, get: function () { return assets_1.getAssetDirectory; } });
Object.defineProperty(exports, "hasGuestAssets", { enumerable: true, get: function () { return assets_1.hasGuestAssets; } });
