// @earendil-works/gondolin/dist/src/sandbox-controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandboxController = void 0;
const events_1 = require("events");
const child_process_1 = require("child_process");
class SandboxController extends events_1.EventEmitter {
    constructor(config) {
        super();
        this.config = config;
        this.child = null;
        this.state = "stopped";
        this.restartTimer = null;
        this.manualStop = false;
    }
    setAppend(append) {
        this.config.append = append;
    }
    getState() {
        return this.state;
    }
    async start() {
        if (this.child)
            return;
        this.manualStop = false;
        this.setState("starting");
        const args = buildQemuArgs(this.config);
        this.child = (0, child_process_1.spawn)(this.config.qemuPath, args, {
            stdio: ["ignore", "pipe", "pipe"],
        });
        this.child.stdout?.on("data", (chunk) => {
            this.emit("log", chunk.toString());
        });
        this.child.stderr?.on("data", (chunk) => {
            this.emit("log", chunk.toString());
        });
        this.child.on("spawn", () => {
            this.setState("running");
        });
        this.child.on("error", (err) => {
            this.child = null;
            this.setState("stopped");
            this.emit("exit", { code: null, signal: null, error: err });
        });
        this.child.on("exit", (code, signal) => {
            this.child = null;
            this.setState("stopped");
            this.emit("exit", { code, signal });
            if (this.manualStop) {
                this.manualStop = false;
                return;
            }
            if (this.config.autoRestart) {
                this.scheduleRestart();
            }
        });
    }
    async stop() {
        if (!this.child)
            return;
        const child = this.child;
        this.child = null;
        this.manualStop = true;
        if (this.restartTimer) {
            clearTimeout(this.restartTimer);
            this.restartTimer = null;
        }
        child.kill("SIGTERM");
        await new Promise((resolve) => {
            const timeout = setTimeout(() => {
                child.kill("SIGKILL");
            }, 3000);
            child.once("exit", () => {
                clearTimeout(timeout);
                resolve();
            });
        });
        this.setState("stopped");
    }
    async restart() {
        await this.stop();
        await this.start();
    }
    scheduleRestart() {
        if (this.restartTimer)
            return;
        this.restartTimer = setTimeout(() => {
            this.restartTimer = null;
            void this.start();
        }, 1000);
    }
    setState(state) {
        if (this.state === state)
            return;
        this.state = state;
        this.emit("state", state);
    }
}
exports.SandboxController = SandboxController;
function buildQemuArgs(config) {
    const args = [
        "-nodefaults",
        "-no-reboot",
        "-m",
        config.memory,
        "-smp",
        String(config.cpus),
        "-kernel",
        config.kernelPath,
        "-initrd",
        config.initrdPath,
        "-append",
        config.append,
        "-nographic",
    ];
    if (config.rootfsPath) {
        args.push("-drive", `file=${config.rootfsPath},format=raw,if=none,id=drive0,snapshot=on`);
        args.push("-device", "virtio-blk-pci,drive=drive0");
    }
    const targetArch = detectTargetArch(config);
    const machineType = config.machineType ?? selectMachineType(targetArch);
    args.push("-machine", machineType);
    const accel = config.accel ?? selectAccel();
    if (accel)
        args.push("-accel", accel);
    const cpu = config.cpu ?? selectCpu();
    if (cpu)
        args.push("-cpu", cpu);
    if (config.console === "none") {
        args.push("-serial", "none");
    }
    else {
        args.push("-serial", "stdio");
    }
    args.push("-object", "rng-random,filename=/dev/urandom,id=rng0");
    args.push("-device", "virtio-rng-pci,rng=rng0");
    args.push("-chardev", `socket,id=virtiocon0,path=${config.virtioSocketPath},server=off`);
    args.push("-chardev", `socket,id=virtiofs0,path=${config.virtioFsSocketPath},server=off`);
    args.push("-device", "virtio-serial-pci,id=virtio-serial0");
    args.push("-device", "virtserialport,chardev=virtiocon0,name=virtio-port,bus=virtio-serial0.0");
    args.push("-device", "virtserialport,chardev=virtiofs0,name=virtio-fs,bus=virtio-serial0.0");
    if (config.netSocketPath) {
        args.push("-netdev", `stream,id=net0,server=off,addr.type=unix,addr.path=${config.netSocketPath}`);
        const mac = config.netMac ?? "02:00:00:00:00:01";
        args.push("-device", `virtio-net-pci,netdev=net0,mac=${mac}`);
    }
    return args;
}
function detectTargetArch(config) {
    const qemuPath = config.qemuPath.toLowerCase();
    if (qemuPath.includes("aarch64") || qemuPath.includes("arm64")) {
        return "arm64";
    }
    if (qemuPath.includes("x86_64") || qemuPath.includes("x64")) {
        return "x64";
    }
    return process.arch;
}
function selectMachineType(targetArch) {
    if (process.platform === "linux" && targetArch === "x64") {
        return "microvm";
    }
    if (targetArch === "arm64") {
        return "virt";
    }
    return "q35";
}
function selectAccel() {
    if (process.platform === "linux")
        return "kvm";
    if (process.platform === "darwin")
        return "hvf";
    return "tcg";
}
function selectCpu() {
    if (process.platform === "linux" || process.platform === "darwin") {
        return "host";
    }
    return "max";
}
